ldap_sync/exporter.py

```diff
@@ -40,35 +57,65 @@ class LdapExporter(object):
 
-        for l, record in enumerate(desired):
+        for l, record in enumerate(desired, 1):
             self.states_dict[record.dn].desired = record
-        logger.info("Gathered %d records lof desired state", l)
+        logger.info("Gathered %d records of desired state", l)
 
         self.actions = []
 
     @classmethod
-    def from_orm_objects_and_ldap_result(cls, current, desired, base_dn):
+    def from_orm_objects_and_ldap_result(
+            cls, ldap_users: dict, db_users: Iterable[UserProxyType], user_base_dn,
+            ldap_groups: dict = None, db_groups: Iterable[GroupProxyType] = None,
+            group_base_dn=None, ldap_properties: dict = None,
+            db_properties: Iterable[PropertyProxyType] = None, property_base_dn=None):
         """Construct an exporter instance with non-raw parameters
 
-        :param cls:
-        :param current: An iterable of records as returned by
-            ldapsearch
-        :param desired: An iterable of sqlalchemy result proxies as returned
-            by :py:func:`fetch_users_to_sync`
+        :param ldap_users: An iterable of records as returned by :func:`fetch_current_ldap_users`.
+        :param db_users: An iterable of sqlalchemy result proxies as returned by :func:`fetch_users_to_sync`.
+        :param user_base_dn:
+        :param ldap_groups: An iterable of records as returned by :func:`fetch_current_ldap_groups`.
+        :param db_groups: An iterable of sqlalchemy result proxies as returned by :func:`fetch_groups_to_sync`.
+        :param group_base_dn:
+        :param ldap_properties: An iterable of records as returned by :py:func:`fetch_current_ldap_properties`.
+        :param db_properties: An iterable of sqlalchemy result proxies as returned by :func:`fetch_properties_to_sync`.
+        :param property_base_dn:
+        :return:
         """
-        return cls((cls.Record.from_ldap_record(x) for x in current),
-                   (cls.Record.from_db_user(x.User, base_dn, x.should_be_blocked)
-                    for x in desired))
+        current = []
+        current.append(UserRecord.from_ldap_record(x) for x in ldap_users)
+        if ldap_groups:
+            current.append(GroupRecord.from_ldap_record(x) for x in ldap_groups)
+        if ldap_properties:
+            current.append(GroupRecord.from_ldap_record(x) for x in ldap_properties)
+
+
+        desired = []
+        desired.append(UserRecord.from_db_user(x.User, user_base_dn, x.should_be_blocked)
+                    for x in db_users)
+
+        # Remove members that are not exported from groups/properties
+        exported_users = { u.User.login for u in db_users }
+
+        if db_groups:
+            desired.append(
+                GroupRecord.from_db_group(x.Group.name,
+                                          (m for m in x.members if m in exported_users),
+                                          group_base_dn, user_base_dn) for x in db_groups)
```

Comment:
Please indent the `for x in db_groups` part of the genexp on a same level as `GroupRecord.[…]`, because at the moment it may seem to be a genexp inside the `from_db_group()` call.